modules/ding_gatewayf/ding_gatewayf.module

```diff
@@ -0,0 +1,458 @@
+      // prefixed with gatewayf, but just wayf so it can be used by all wayf
+      // modules to indicate that the user was logged in via wayf. This could
+      // then be used to change behaviour in forgotten pin-code in the user
+      // profile (not requiring old password when logged in with wayf).
+      $_SESSION[DING_GATEWAYF_IS_LOGGED_IN] = TRUE;
+
+      // Gatewayf does some url encoding of the destination parameter in the
+      // request, so we need to clean it up.
+      if (!empty($_REQUEST['destination'])) {
+        $_REQUEST['destination'] = substr($_REQUEST['destination'], 0, strpos($_REQUEST['destination'], '?'));
+      }
+
+      // Log into the library system and Drupal.
+      _ding_gatewayf_provider_login($attributes);
+    }
+    else {
+      watchdog('ding_gatewayf', 'The gateway returned no error and no eduPersonTargetedID', array(), WATCHDOG_ERROR);
+
+      // Set message to the user.
+      drupal_set_message(t('The login with nem-login failed please try again later.'), 'error');
+
+      _ding_gatewayf_redirect_user('<front>');
+    }
+  }
+  else {
+    // Log the user out of drupal an clear session.
+    unset($_SESSION[DING_GATEWAYF_IS_LOGGED_IN]);
+
+    // The gatewayf logout service does not return any information about the
+    // success of the logout request, so the only thing we can do is redirect to
+    // drupal logout.
+    _ding_gatewayf_redirect_user('user/logout');
+  }
+}
+
+/**
+ * Login user to the library system and Drupal.
+ *
+ * @param array $attributes
+ *   Attributes returned by gatewayf service.
+ */
+function _ding_gatewayf_provider_login(array $attributes) {
+  if (ding_provider_implements('wayf', 'login_credentials')) {
+    // Parse the attributes required by the current provider.
+    $required_attr = _ding_gatewayf_get_credentials($attributes);
+
+    // Ding gatewayf comes with an hidden variable that can be used to map the
+    // CPR return from gatewayf to another user. This is very useful during
+    // development, if the developer does not have real library account.
+    if (variable_get('ding_gatewayf_cpr_mapping', 0)) {
+      // It's inserted into the name credential.
+      $required_attr['schacPersonalUniqueID'] = variable_get('ding_gatewayf_cpr_mapping', '');
+    }
+
+    // Get credentials from the provider.
+    $credentials = ding_provider_invoke('wayf', 'login_credentials', $required_attr);
+    if (!empty($credentials)) {
+      // Set values to send with the login form.
+      $form_state = array(
+        'values' => $credentials,
+      );
+
+      // Submit login form to trigger login, where the login process goes
+      // through the whole process in ding_user.
+      drupal_form_submit('user_login_block', $form_state);
+
+      // Check if the submission trigger any form errors.
+      $errors = form_get_errors();
+      if (!empty($errors)) {
+        if (module_exists('ding_gatewayf_registration')) {
```

Comment:
Adding a comment explaining that is *why* we check for errors would be good then.